compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1903,6 +1905,28 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         // Extract the match-pair from the highest priority candidate and build a test from it.
         let (match_place, test) = self.pick_test(candidates);
 
+        // Patterns like `Some(A) | Some(B)` need spans for their target branches separately.
+        // Keep it in original order so that true arms of the laters won't be counted as false arms of the formers.
+        let mut mcdc_coverage_targets: Option<Vec<_>> =
+            self.tcx.sess.instrument_coverage_mcdc().then(|| {
+                candidates
+                    .iter()
+                    .filter_map(|candidate| {
+                        candidate.match_pairs.first().map(|matched_pair| {
```

Comment:
The first match pair likely does not correspond to the match pair that was matched on (except for the first candidate in the list). E.g.
```rust
match (0, true) {
  (_, true) => ... // chosen match pair is `scrutinee.1 @ true`
  (0, false) => ... // chosen match pair is `scrutinee.0 @ 0`
}
```

The reliable way would be to return the span you need in the `target_candidates` output of `sort_candidates`

Reply:
Tried this but `target_candidates` might lose the very first test.
For example, suppose the pattern is `if let (A | B, C) = val`, I want to record the span of test `C` and `A | B`. In `sort_candidates`, `C` is `full_matched` and is removed from `candidates`. Therefore span of `C` cannot be found in `target_candidates`.
Actually I want to record spans of all candidates and their "matched blocks" (excluding the whole). Could I get them here?

Reply:
Yeah, I suggest that whenever a match pair is `fully_matched`, you keep its span. So `target_candidates` would contain `(Vec<Candidate>, Span)`.

Also, in your example the span of `A | B` is lost before that because we expand or-patterns in `match_candidates`. You can only get the spans of `A` and `B` on their own.

Reply:
Oh sorry I had made it ambiguous
Let's consider such code 
```rust
// Suppose val: (Pat, Pat)
match val {
  (Pat::A | Pat::B, Pat::D) => { /*...*/ }
  (Pat::C, Pat::E) => { /*...*/ }
  _ => { /*...*/ }
}
```
First we check val.1. Here we have two target candidates: `Pat::D` and `Pat::E`. The `test_block` means which block the discriminant is checked, in this function it's `start_block`. And if the discriminant shows the pattern is `Pat::D`, then we switch to a block, suppose `bb2`, where we can extract value inside `Pat::D` and tests remains. In this way I mean `bb2` is the matched block for `Pat::D`, in this function it's `candidate_start` in iteration over `target_candidates`.

The question is, when we first call `test_candidates` here, we have two target candidates, whose spans are `(Pat::A | Pat::B, Pat::D)` and `(Pat::C, Pat::E)` respectively. How could I shrink them to `Pat::D` and `Pat::E` respectively?

Reply:
> Yeah, I suggest that whenever a match pair is `fully_matched`, you keep its span. So `target_candidates` would contain `(Vec<Candidate>, Span)`.
> 

Is this right? Only keep the candidate span and it's `full_matched` pair so that we can find what is matched later. (Just a draft here, might need to pass a argument to `sort_candidates` in final version).
```rust
        // Record all candidates and their matched pairs
        let mut candidate_matched_spans = self.tcx.sess.instrument_coverage_mcdc().then(|| {
            candidates
                .iter()
                .map(|candidate| {
                    (
                        candidate.extra_data.span,
                        candidate
                            .match_pairs
                            .iter()
                            .map(|matched_pair| matched_pair.pattern.span)
                            .collect::<std::collections::BTreeSet<_>>(),
                    )
                })
                .collect::<std::collections::BTreeMap<_, _>>()
        });

        // For each of the N possible test outcomes, build the vector of candidates that applies if
        // the test has that particular outcome.
        let (remaining_candidates, target_candidates) =
            self.sort_candidates(match_place, &test, candidates);

        // Find full matched pairs
        let matched_candidates = candidate_matched_spans.map(|mut matched| {
            for (_, targets) in target_candidates.iter() {
                for tgt in targets {
                    let Some(full_matched_pairs) = matched.get_mut(&tgt.extra_data.span) else {
                        continue;
                    };
                    for matched_pair in &tgt.match_pairs {
                        full_matched_pairs.remove(&matched_pair.pattern.span);
                    }
                    assert_eq!(full_matched_pairs.len(), 1);
                }
            }
            matched
                .into_iter()
                .filter_map(|(candidate_span, full_matched_pairs)| {
                    (full_matched_pairs.len() == 1).then_some((
                        candidate_span,
                        (full_matched_pairs.into_iter().next().expect("checked"), None),
                    ))
                })
                .collect::<std::collections::BTreeMap<_, (Span, Option<BasicBlock>)>>()
        });
```

Reply:
> The question is, when we first call test_candidates here, we have two target candidates, whose spans are (Pat::A | Pat::B, Pat::D) and (Pat::C, Pat::E) respectively. How could we shrink them to Pat::D and Pat::E respectively?

You should not use the candidate spans, use the match_pair spans instead, these are the spans you're looking for.